OpenRA.Mods.Common/UpdateRules/Rules/20191117/ReplaceAttackTurnDelay.cs

```diff
@@ -0,0 +1,67 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class ReplaceAttackTurnDelay : UpdateRule
+	{
+		public override string Name { get { return "Removed AttackTurnDelay from AttackAircraft and replaced with AttackRunLength."; } }
+		public override string Description
+		{
+			get
+			{
+				return "Removed AttackTurnDelay which defines the strafing run in ticks from AttackAircraft and replaced it with AttackRunLength"
+				+ " which defines the strafing run in WDist distance.";
+			}
+		}
+
+		readonly List<Tuple<string, string>> strafers = new List<Tuple<string, string>>();
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			var message = "AttackTurnDelay is removed from AttackAircraft. Attack run lengths are now calculated dynamically.\n"
+				+ "You may want to manually override the lengths in the following places by defining AttackRunLength:\n"
+				+ UpdateUtils.FormatMessageList(strafers.Select(n => n.Item1 + " (" + n.Item2 + ")"));
+
+			if (strafers.Any())
+				yield return message;
+
+			strafers.Clear();
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var aircraft = actorNode.LastChildMatching("Aircraft");
+			if (aircraft != null)
+			{
+				var attackAircraft = actorNode.LastChildMatching("AttackAircraft");
+				if (attackAircraft == null)
+					yield break;
+
+				var canHover = aircraft.LastChildMatching("CanHover");
+				var attackType = attackAircraft.LastChildMatching("AttackType");
+
+				if (canHover != null && canHover.NodeValue<bool>() == true && attackType != null && attackType.NodeValue<AirAttackType>() != AirAttackType.Strafe)
```

Comment:
In practice this check won't do much because the `CanHover` is typically inherited from a default template.
We support hovering units that strafe, so i don't think its valid to throw those out here anyway?

In that case, we can simplify down to:
```csharp
if (attackType != null && attackType.NodeValue<AirAttackType>() != AirAttackType.Strafe)
```

which fixes the false-positive reporting of the RA and TD helicopters.

Reply:
We support hovering units that strafe, but we don't support non-hovering units that don't strafe, which is what is excluded here.

Reply:
The issue here is that we are reporting units with `AirAttackType.Hover` in the `StrafeRunLength` notification, even though `StrafeRunLength` doesn't apply to them.

Reply:
Yes, but otherwise we would theoretically skip units for which `StrafeRunLength` does apply. The issue is with FlyAttack.cs: L157. Even when the aircraft explicitly defines another `AttackType` it will still do a strafing run if it cannot hover. This is pretty counterintuitive though, so perhaps we should change that line and just allow the nonsensical behaviour. After all: `AttackType` defaults to `Strafe` anyway.

Reply:
My main motivation is to remove the `Aircraft` trait lookup because these kind of lookups don't reliably work from update rules. IMO its not worth the effort to notify the user that they can customize an already-broken configuration, which lead to my suggestion above.